120 continuous integration

lylaminju opened the pull request:
## Summary

<!-- Describe what problem you are tying to solve -->

- Added a CI file that installs dependencies and run tests
- The CI is triggered when code is pushed to the `main` or `dev` branches, or when a pull request is opened targeting those branches.

## Tests

<!-- Tell us if the changes have been tested and how -->

## Additional information

<!-- Do we need additional context or information about the changes? Please describe it here -->

- Although `Husky` handles linting locally, I included a lint check in the CI pipeline to provide a more reliable double-check.

## References
https://github.com/Orange-OpenSource/hurl?tab=readme-ov-file#debian--ubuntu

lylaminju committed:
fix: add missing property in [assets] section

lylaminju committed:
chore: skip failing integration tests

lylaminju committed:
ci: add Github Actions workflow for test

lylaminju committed:
ci: add lint check

lylaminju committed:
ci: split lint and test, add integration test

manvinderjit commented:
> Just my suggestions, @madcampos and @kbventures might have more insights on this.

nitpick: The repo already has some bash script files in the `.husky` folder. We can leverage them in the yaml file instead of writing commands. 
We can also create bash files to segregate the commands for different purposes, like put all the integration testing commands in a single integration-tests.sh bash file.

madcampos commented on .github/ci.yml:
```diff
@@ -0,0 +1,82 @@
+      - name: Run unit tests
+        run: npm test
+
+      - name: Start preview server (background)
+        run: |
+          npm run build
+          npx vite preview --port 3000 &
+          echo $! > preview.pid
+          sleep 3
+
+      - name: Install Hurl CLI
+        run: |
+          VERSION=6.1.1
+          curl --location --remote-name https://github.com/Orange-OpenSource/hurl/releases/download/$VERSION/hurl_${VERSION}_amd64.deb
+          sudo apt update
+          sudo apt install -y ./hurl_${VERSION}_amd64.deb
+
+      - name: Run integration tests
+        run: |
+          set -e
+          find . -name "*.hurl" | while read file;
+          do
+            echo "Running: $file";
+            hurl "$file";
+          done
+
+      - name: Kill preview server
+        if: always()
+        run: |
+          if [ -f preview.pid ]; then
+            kill $(cat preview.pid)
+          fi
```
Let's put this on a different job, so each type of testing runs on it's own and if things fail we can re-run only one thing like unit tests instead of running everything.

Another question: can we cache/reuse steps? Like, the step for installing node and installing dependencies is repeated multiple times, could we cache that?

lylaminju replied:
> Another question: can we cache/reuse steps? Like, the step for installing node and installing dependencies is repeated multiple times, could we cache that?

Yes, great observation! I'm already using `cache: "npm"` in the setup-node action, which caches npm dependencies between runs. 

To reduce the duplicate setup code across jobs, I'll create a [Composite Action](https://docs.github.com/en/actions/sharing-automations/creating-actions/creating-a-composite-action?platform=mac) that bundles the checkout, node setup, and npm install steps. This will make the workflow much cleaner and more maintainable.

madcampos commented on wrangler.toml:
```diff
@@ -19,6 +19,7 @@ enabled = true
 head_sampling_rate = 1
 
 [assets]
+directory = "public"
```
"Oops..." I was trigger happy and remove this line, sorry.

lylaminju committed:
ci: separate unit tests and integration tests

so that each type of testing runs on it's own and if things fail we can re-run only one thing instead of running everything

lylaminju committed:
ci: utilized Composite Action to improve ci.yml

madcampos commented:
> > Just my suggestions, @madcampos and @kbventures might have more insights on this.
> 
> nitpick: The repo already has some bash script files in the `.husky` folder. We can leverage them in the yaml file instead of writing commands. We can also create bash files to segregate the commands for different purposes, like put all the integration testing commands in a single integration-tests.sh bash file.

I'm going to be very annoying here...

The bash scripts on the `.husky` folder are for git hooks, even if there is some duplication, we should not mess with them. They basically just call something when we do something on git.

What we can do however is update the `scripts` section on the `package.json` file to include whatever script we would like, that is better and it also allows anyone to run the scripts independently.

On adding shell scripts I will say the same thing I said to Ken before: no. 🙃

I would like to avoid shell scripts in particular because they are not portable and don't work well across OSes.

I think for most direct things we can use `npm run ...` and set the scripts we want. Note that every script can also have two other attached scripts to it: `pre<name>` and `post<name>`, and they can run multiple commands using `&&` (this works cross OS).

If we need more complex setups, that may be the case for writing build scripts, on that case, we should write them in js so they can run with node and we don't introduce a new dependency, and we can add them to the `build` folder.

kbventures reviewed (changes requested):
@lylaminju 

The scripts will need include to include  {{host}} variable for the hurl scripts to work. 

Currently we are passing the host value using a terminal command paramater "--variable host=localhost:3000"

```
set -e
find . -name "*.hurl" | while read file;
do
  echo "Running: $file";
  hurl --variable host=http://localhost:3000 "$file";
done
```

lylaminju committed:
ci: use hurl globbing for integration tests instead of for-loop

lylaminju commented:
@kbventures @madcampos Thanks for the suggestions!  
Since we only have a single `host` variable for now, I opted not to create a `.env` file.
If we add more variables in the future, setting up an environment file would definitely make things cleaner.

lylaminju committed:
Merge branch 'dev' into 120-continuous-integration
